starknet/src/tests/test_upgrade.cairo

```diff
@@ -0,0 +1,128 @@
+    use debug::PrintTrait;
+    use serde::{Serde};
+
+    use sx::space::space::{Space, ISpaceDispatcher, ISpaceDispatcherTrait};
+    use sx::authenticators::vanilla::{
+        VanillaAuthenticator, IVanillaAuthenticatorDispatcher, IVanillaAuthenticatorDispatcherTrait
+    };
+    use sx::execution_strategies::vanilla::VanillaExecutionStrategy;
+    use sx::voting_strategies::vanilla::VanillaVotingStrategy;
+    use sx::proposal_validation_strategies::vanilla::VanillaProposalValidationStrategy;
+    use sx::tests::mocks::proposal_validation_always_fail::AlwaysFailProposalValidationStrategy;
+    use sx::utils::types::{
+        Strategy, IndexedStrategy, Choice, FinalizationStatus, Proposal, UpdateSettingsCalldataImpl
+    };
+    use sx::utils::constants::{PROPOSE_SELECTOR, VOTE_SELECTOR, UPDATE_PROPOSAL_SELECTOR};
+    use sx::tests::setup::setup::setup::{setup, deploy};
+    use sx::interfaces::{
+        IProposalValidationStrategyDispatcher, IProposalValidationStrategyDispatcherTrait
+    };
+    use sx::tests::mocks::executor::{
+        ExecutorExecutionStrategy, ExecutorExecutionStrategy::Transaction
+    };
+    use starknet::ClassHash;
+
+    use Space::Space as SpaceImpl;
+
+    #[test]
+    #[available_gas(10000000000)]
+    fn test_upgrade() {
+        let config = setup();
+        let (factory, space) = deploy(@config);
+
+        // New implementation is not a proposer space but a random contract (here, a proposal validation strategy).
+        let new_implem = AlwaysFailProposalValidationStrategy::TEST_CLASS_HASH.try_into().unwrap();
+
+        testing::set_contract_address(config.owner);
+
+        // Now upgrade the implementation
+        space.upgrade(new_implem);
+
+        // Ensure it works
+        let new_space = IProposalValidationStrategyDispatcher {
+            contract_address: space.contract_address
+        };
+
+        let author = contract_address_const::<0x7777777777>();
+        let params = ArrayTrait::<felt252>::new();
+        let user_params = ArrayTrait::<felt252>::new();
+        let res = new_space.validate(author, params, user_params);
+        assert(res == false, 'Strategy did not return false');
+    }
+
+    #[test]
+    #[available_gas(10000000000)]
+    fn test_upgrade_via_execution_strategy() {
+        let config = setup();
+        let (factory, space) = deploy(@config);
+        let proposal_id = space.next_proposal_id();
+
+        // New implementation is not a proposer space but a random contract (here, a proposal validation strategy).
+        let new_implem: ClassHash = AlwaysFailProposalValidationStrategy::TEST_CLASS_HASH
+            .try_into()
+            .unwrap();
+
+        let (execution_contract_address, _) = deploy_syscall(
+            ExecutorExecutionStrategy::TEST_CLASS_HASH.try_into().unwrap(),
+            0,
+            ArrayTrait::<felt252>::new().span(),
+            false
+        )
+            .unwrap();
+
+        // Set the owner to be the space itself
```

Comment:
Set the owner to be the execution strategy

Reply:
386c9484ad3e51f6a04354e9c523b10805fd7daa